### ISSUE TITLE
mix and task parameters

### ISSUE THREAD

**Issue by emmanueltouzery** [stevearc/overseer.nvim#71]
I noticed now that although overseer now correctly detects elixir mix tasks, i doesn't know which of these may require parameters.
Sadly it seems that's just metadata that mix doesn't advertise...

For instance:
https://github.com/elixir-ecto/ecto_sql/blob/v3.9.1/lib/mix/tasks/ecto.gen.migration.ex#L91

There is some metadata (not all of which may be accessible through the command-line), but the fact that a command-line parameter is required is not advertised anywhere, instead we get a failure with a human-readable, not computer-readable error message.

There are multiple solutions here I guess:
1. not register the mix tasks automatically but manually, where the user has complete control
2. allow the user to "patch" tasks -- with some configuration for overseer maybe, eg mix task name->parameter list for that task
3. ugly workaround

Thinking about 2, I was looking at the overseer.wrap_template() call in mix.lua in the overseer repo. If there we added some params=... where we would for instance call some optional function that the user could provide, this could work, possibly.
But for now I don't really understand overseer params..
I tried to edit the mix.lua like that:
```lua
            overseer.wrap_template(
              tmpl,
              {
                name = string.format("mix %s", task_name),
                params = {
                  my_var = {
                    type = "string",
                    name = "More readable name",
                    desc = "A detailed description",
                    optional = false,
                  },
                  myvar2 = {
                    type = "string",
                    name = "More readable name",
                    desc = "A detailed description",
                    optional = false,
                  } }
                },
              { args = { task_name } }
            )
```

But nothing changed in the behavior, OverseerRun doesn't ask me to enter the parameters, so I'm obviously doing something wrong.

Another option is a horrible hack :) which is what I have now :)

```lua
    -- add other tasks here
    mix_params = {{"ecto.gen.migration", "Enter migration name: "}}

    require('overseer').add_template_hook({name="mix"}, function(task_defn, util)
      for idx, km in ipairs(mix_params) do
        if task_defn.cmd[2] == km[1] then
          task_defn.args = {vim.fn.input(km[2])}
        end
      end
    end)
```

actually i would be happy if i could using vim.ui.input in there, which looks nicer than vim.fn.input, but due to the async, I couldn't make it work. But it's clear to me that this approach is a non-starter for you.

Do you think something could be done for this use case in overseer? and it's very clear to me that the flaw is in mix, not in overseer. And approach 1 may just be the correct one.

**Comment by stevearc** [comment]
I don't see how doing option 2 or 3 would be better than simply defining the task yourself. Either way, you have to put in a specific customization for each task you care about.

I know nothing about elixir or mix, but would it be possible to extract argument information and surface it either in the human-readable `mix help` or in a new machine-readable command? If so, maybe make an issue or a pull request against mix to add that functionality. Seems like it would be pretty broadly useful.

**Comment by emmanueltouzery** [comment]
Yes... Maybe it could be possible to at least select the tasks which are parameterless, and keep those, and hide the others "must be configured by hand". But I'm not sure how best to achieve that.

Changing mix.. would be the best for sure but I'm unlikely to make it.. and even when done, it'll take quite some time before I can use the latest version and libraries upgrade and so on. But it would be the best way for sure.

So now I'm getting some detected tasks which are misconfigured and wondering how to give them. In the end maybe removing the mix support is the most reasonable action :(

**Comment by stevearc** [comment]
Unfortunately I don't have any great solutions for you. It's hard to do something programmatically when you don't have enough data. I really would encourage you to at least request this upstream, though. It may not be fast, but a good solution eventually is better than no solution ever.

**Comment by emmanueltouzery** [comment]
I spent some more time thinking about this. I also wrote a draft feature request for the elixir mix project but... I don't think they would accept it or that it would make sense for them, so I didn't send it in the end. I believe that their starting point is that the parameters for these tasks are completely turing complete at this point. They could do anything. They support a series of flags, and a series of optional features. For instance the `mix test` task examples support things like that:
```
mix test --exclude test --include line:12 --include line:24 test/some/particular/file_test.exs
```

And even more things like flags to enable test coverage and so on. So I think they'd conclude that exposing a complete programmatically discoverable API for the tasks would be overkill and too much work for very little added value.

One option that I can see would be to always open a popup asking you for the task parameters. I could set `default_template_prompt=always` in my overseer setup and it would fix the issue for me, but then it would apply also to my other projects which have for instance npm tasks that don't have this issue and so on.

A solution would be to be able to specify in a template that this specific template requires a mandatory prompt for parameters. Then we would set this up for the mix support. That way users can always run mix tasks, even if they require parameters.

Then ideally personally, i'd really like to be able to manually override for a couple of tasks to say "well this one actually doesn't need the prompt". But that would be a bonus...

Besides that I don't really see what else to do. Unfortunately this idea adds more complexity in overseer :( so i would totally understand if that doesn't make sense for you. However in its current state the mix support is flawed, so if nothing is changed, I wouldn't exclude removing it.
